Implements TrueBeats data source

gsalaz98 opened the pull request:
Tests are a WIP at the moment, but this is the first of 4 data sets that will be implemented for ExtractAlpha. This data set also happens to be the most complicated one of the 4 to process, so the subsequent implementations should be faster and easier to implement

gsalaz98 committed:
Initial implementation of ExtractAlpha TrueBeats data source

gsalaz98 committed:
Clean up and document code, sync script improvements, bug fixes

  * Beginning work on tests

gsalaz98 committed:
Add support for processing historical data (WIP)

  * Reprocessing non-historical data (2021-08-31) produced 100% identical
    results as the previous commit (verified by checking MD5 hash of all
    outputted files with MD5 of previous files)

gsalaz98 committed:
Improve historical data processing

Martin-Molinero commented on DataProcessing/ExtractAlphaTrueBeatsHistoricalConverter.cs:
```diff
@@ -0,0 +1,258 @@
+                {
+                    foreach (var trueBeat in trueBeats)
+                    {
+                        // Let's reset any values we previously set back to the default
```
Why would we need to do this? 🤔

Martin-Molinero commented on DataProcessing/ExtractAlphaTrueBeatsHistoricalConverter.cs:
```diff
@@ -0,0 +1,258 @@
+            ConvertHistoricalData();
+            
+            var startDate = new DateTime(2002, 1, 1);
+            var endDate = new DateTime(2021, 1, 31);
```
Let's make these values a config with default value 👍

Martin-Molinero commented on DataProcessing/Program.cs:
```diff
@@ -0,0 +1,33 @@
+            var outputDataDirectory = Directory.CreateDirectory(Config.Get("temp-output-directory", "/temp-output-directory"));
+
+            var converter = processHistoricalData 
+                ? new ExtractAlphaTrueBeatsHistoricalConverter(deploymentDate, rawDataDirectory, existingDataDirectory, outputDataDirectory)
```
How will `deploymentDate` affect the `ExtractAlphaTrueBeatsHistoricalConverter` behavior? I wouldn't expect it to be an argument, it's fishy, can we remove it?

Martin-Molinero commented on ExtractAlphaTrueBeat.cs:
```diff
@@ -0,0 +1,215 @@
+                ManagementBeat = managementBeat,
+                
+                Time = date,
+                EndTime = date.Date.AddHours(12).AddMinutes(30),
```
Why is `date.Date.AddHours(12).AddMinutes(30)` the endtime?

Martin-Molinero commented on ExtractAlphaTrueBeat.cs:
```diff
@@ -0,0 +1,215 @@
+
+using System;
```
nit - think all the files are missing the license header

Martin-Molinero commented on QuantConnect.DataSource.csproj:
```diff
@@ -3,13 +3,13 @@
+    <PackageReference Include="QuantConnect.Common" Version="2.5.*" />
     <PackageReference Include="protobuf-net" Version="3.0.29" />
-    <PackageReference Include="protobuf-net.Core" Version="3.0.29" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
```
`Newtonsoft.Json` isn't used I think?

Martin-Molinero commented on tests/Tests.csproj:
```diff
@@ -3,9 +3,6 @@
     <RootNamespace>QuantConnect.DataLibrary.Tests</RootNamespace>
   </PropertyGroup>
-  <ItemGroup>
-    <Compile Include="..\Demonstration.cs" Link="Demonstration.cs" />
```
Demonstration should be updated and included I think?

Martin-Molinero commented on DataProcessing/ExtractAlphaTrueBeatsConverter.cs:
```diff
@@ -0,0 +1,460 @@
+                            // and the TrueBeat calculation fluctuates between two numbers.
+                            if (_duplicateFQ1DataTickers.Add(ticker))
+                            {
+                                Log.Error($"ExtractAlphaTrueBeatsConverter.ParseTrueBeats(): Duplicate data encountered in FQ1 dataset for ticker: {ticker} - skipping");
```
Could these be updates emitted at different times? Makes me wonder if the Time of the baseData being used is the right one

gsalaz98 replied:
I don't think so, there's no way to know which event came first, and the fact that data is duplicated on multiple rows is suspicious since almost every single ticker in that data doesn't have duplicates

Martin-Molinero commented on DataProcessing/ExtractAlphaTrueBeatsConverter.cs:
```diff
@@ -0,0 +1,460 @@
+                    $"{ticker.ToLowerInvariant()}.csv");
+                
+                Log.Trace($"ExtractAlphaTrueBeatsConverter.WriteToFile(): Writing data to: {outputFilePath}");
+                File.WriteAllText(outputFilePath, string.Join("\n", outputDataLines));
```
Hm `outputDataLines` being a hash set doesn't guarantee the order, don't think this is right? Should probably just override existing data here?

Martin-Molinero commented on DataProcessing/ExtractAlphaTrueBeatsConverter.cs:
```diff
@@ -0,0 +1,460 @@
+                        }
+
+                        var (fiscalYear, fiscalQuarter) = ParseFiscalPeriod(csv[_fiscalPeriodIndex]);
+                        if (!allQuarters && symbolTrueBeats.Any(x =>
+                            x.Time == processingDate &&
+                            x.EarningsMetric == earningsMetric &&
+                            x.FiscalPeriod.FiscalYear == fiscalYear &&
+                            x.FiscalPeriod.FiscalQuarter == fiscalQuarter &&
+                            x.ExpertBeat != null &&
+                            x.TrendBeat != null &&
+                            x.ManagementBeat != null))
```
This compare logic should probably be owned by the data type itself,

Martin-Molinero commented on DataProcessing/ExtractAlphaTrueBeatsConverter.cs:
```diff
@@ -0,0 +1,460 @@
+                    if (_duplicateFiscalDataTickers.Add(ticker))
+                    {
+                        Log.Error($"ExtractAlphaTrueBeatsConverter.ParseTrueBeats(): Duplicate data encountered in Fiscal_Periods_EPSSales dataset for ticker: {ticker} - skipping");
+                    }
```
Similar comments here 👍

Martin-Molinero reviewed (commented):
Cool! Leaving a couple of comments 👍 

*CI is broken BTW

gsalaz98 committed:
Address review (partial): Data sorting bug fix, code cleanup

  * Add comments to code clarifying the intention behind decisions
  * Add license header to files in the project
  * Remove unused dependency

gsalaz98 commented:
I've addressed the review partially, pending some feedback on a few concerns of yours @Martin-Molinero 

I'll add the demo algo once I've finished tests as well, since it's the least critical component of the implementation at the moment.

gsalaz98 committed:
Address review: finish adding tests, log improvement, and CI setup

gsalaz98 committed:
Bug fix: use time from data instead of date argument in Reader(...)

Martin-Molinero commented on DataProcessing/ExtractAlphaTrueBeatsConverter.cs:
```diff
@@ -349,20 +370,12 @@ protected void WriteToFile(Dictionary<string, List<ExtractAlphaTrueBeat>> proces
                 var outputDataLines = outputData
+                    .OrderBy(x => x.Time)
                     .Select(ToCsv)
+                    .Distinct()
```
Note there is no guarantee of order here see https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.distinct?view=net-5.0#remarks

Martin-Molinero commented on DataProcessing/ExtractAlphaTrueBeatsHistoricalConverter.cs:
```diff
@@ -20,8 +37,8 @@ public class ExtractAlphaTrueBeatsHistoricalConverter : ExtractAlphaTrueBeatsCon
-        public ExtractAlphaTrueBeatsHistoricalConverter(DateTime processingDate, DirectoryInfo rawDataDirectory, DirectoryInfo existingDataDirectory, DirectoryInfo outputDataDataDirectory) 
-            : base(processingDate, rawDataDirectory, existingDataDirectory, outputDataDataDirectory)
+        public ExtractAlphaTrueBeatsHistoricalConverter(DirectoryInfo rawDataDirectory, DirectoryInfo existingDataDirectory, DirectoryInfo outputDataDataDirectory) 
+            : base(DateTime.UtcNow.Date, rawDataDirectory, existingDataDirectory, outputDataDataDirectory)
```
Cool! 👍

Martin-Molinero reviewed (commented):
Cool! Minor comment but important one 😄

gsalaz98 committed:
Address review: ensure sorting is preserved when filtering duplicates

  * Remove unnecessary sorting before writing to file
